### ISSUE TITLE
dirname is not equivalent to the utility

### ISSUE THREAD

**Issue by calimeroteknik** [dylanaraps/pure-bash-bible#76]
Here are a few counter-examples:
```sh
~ $ dirname_pbb() {
    # Usage: dirname "path"
    printf '%s\n' "${1%/*}/"
}

~ $ dirname dir1/dir2/
dir1
~ $ dirname_pbb dir1/dir2/
dir1/dir2

~ $ dirname dir1/dir2
dir1
~ $ dirname_pbb dir1/dir2
dir1/

~ $ dirname dir1
.
~ $ dirname_pbb dir1
dir1/
```
In fact I already tried to reimplement that one with variable substitution and found a few corner cases like this, making me unsure I was up to the task without a serious read of… the POSIX spec, I suppose.

**Comment by dylanaraps** [comment]
Great catch.

This passes your above tests:

```sh
dirname() {
    # Usage: dirname "path"
    dir=${1%%/}

    [[ $dir == */* ]] || dir=.

    printf '%s\n' "${dir%/*}"
}
```

I'll keep digging to see if there are any other cases in which it fails. :+1:

**Comment by dylanaraps** [comment]
I can't reproduce an output in where they differ now. :+1:

**Comment by calimeroteknik** [comment]
The fix for the corner cases I mentioned introduced a regression:
```sh
~ $ dirname_pbb() {
    # Usage: dirname "path"
    dir=${1%%/}

    [[ $dir == */* ]] || dir=.

    printf '%s\n' "${dir%/*}"
}
~ $ dirname_pbb /
.
~ $ dirname /
/

```

**Comment by dylanaraps** [comment]
Working on a fix. Great catch.

**Comment by dylanaraps** [comment]
Fixed.

```sh
dirname() {
    # Usage: dirname "path"
    dir=${1%%/}

    [[ $dir ]] || dir=//
    [[ $dir == */* ]] || dir=.

    printf '%s\n' "${dir%/*}"
}
```

**Comment by calimeroteknik** [comment]
Tests find errors, but unfortunately do not prove correctness:
```sh
~ $ dirname_pbb() {
    # Usage: dirname "path"
    dir=${1%%/}

    [[ $dir ]] || dir=//
    [[ $dir == */* ]] || dir=.

    printf '%s\n' "${dir%/*}"
}
~ $ dirname_pbb /foo

~ $ dirname /foo
/
```
I believe that this would warrant formal verification rather than testing, because it seems tricky.

**Comment by dylanaraps** [comment]
Try this one:

```sh
dirname() {
    # Usage: dirname "path"
    dir=${1%%/}
    dir=${dir%/*}

    [[ $1 == */* ]] || dir=.

    printf '%s\n' "${dir:-/}"
}
```

It passes my ever growing list of testcases. Either way, it's a lot more correct than what was previously in the book. :)

**Comment by calimeroteknik** [comment]
For `something/` it just strips the ending slash rather than returning dot.

**Comment by dylanaraps** [comment]
Here we go again:

```sh
dirname() {
    # Usage: dirname "path"
    dir=${1%%/}
    [[ "${dir##*/*}" ]] && dir=.
    dir=${dir%/*}

    printf '%s\n' "${dir:-/}"
}
```

**Comment by calimeroteknik** [comment]
To still find problems we now need to go into non-normalised paths, for example, `something//`

I'm not even sure that the original `dirname(1)` is actually POSIX-conformant by all accounts anyway, so this might end up being a bug-for-bug reimplementation of the GNU coreutils version: I assume that the standard would be a better reference past this point.

This being said, `dirname(1)` from the GNU coreutils acts as if it normalised all but the prefix slashes; I assume they considered it safer to keep them as-is, seeing as on some systems `//something` is not the same thing as `/something`.

**Comment by dylanaraps** [comment]
I'm testing against `busybox` `dirname` as I don't use `glibc`/GNU `coreutils`. I'm not sure if `busybox` `dirname` is **more** POSIX compliant or not.

I've found another bug to fix though.

```
dirname something//
```

**Comment by calimeroteknik** [comment]
Yes, as noted in my previous comment non-normalised paths are still problematic… but at this point, isn't it safer to just read POSIX rather than retro-engineer busybox through test cases?

The standard describes an algorithm:
https://pubs.opengroup.org/onlinepubs/009695399/utilities/dirname.html

**Comment by dylanaraps** [comment]
Here's an implementation of the algorithm described in the standard:

```sh
dirname_sh() {
    # Usage: dirname "path"
    local tmp=${1:-.}

    [[ $tmp != *[!/]* ]] && {
        printf '/\n'
        return
    }

    tmp=${tmp%%"${tmp##*[!/]}"}

    [[ $tmp != */* ]] && {
        printf '.\n'
        return
    }

    tmp=${tmp%/*}
    tmp=${tmp%%"${tmp##*[!/]}"}

    printf '%s\n' "${tmp:-/}"
}
```

**Comment by dylanaraps** [comment]
```sh
dirname() {
  # Usage: dirname "path"
  local tmp=${1:-.}

  tmp=${tmp%%"${tmp##*[!/]}"}

  [[ ${tmp##*/*} ]] && tmp=.

  tmp=${tmp%/*}
  tmp=${tmp%%"${tmp##*[!/]}"}

  printf '%s\n' "${tmp:-/}"
}
```

**Comment by calimeroteknik** [comment]
The direct implementation looks standard-conformant to me indeed. (beyond reasonable doubt although without machine-assisted checking)

I did not verify the second implementation; unlike `basename` the problem is a little bit too large for me to mentally check that it the second implementation is equivalent to the first. We must recognize our limits as programmers: validating this one without machine assistance is most likely out of my league…

Playing it heuristic-style, if this codegolf passed all tests first time, it's not proof but it's an encouraging sign.

Historical anecdote: hidden files in UNIX were "invented" by accident due to a subtle bug in the, I'd say "codegolf" implementation of `ls`, where it was supposed to ignore the special directories `.` and `..` but in fact ignored all filenames starting in a dot. Oops… But this was (ab?)used and became the standard.

**Comment by dylanaraps** [comment]
I'll push the direct implementation to the bible. :+1:

Regardless of whether or not it is 100% compatible with the standard, this version is a lot "more" compatible with it. At some point I'll set up some proper tests so I can "safely" say "POSIX conformant".

> Historical anecdote: hidden files in UNIX were "invented" by accident due to a subtle bug in the, I'd say "codegolf" implementation of ls, where it was supposed to ignore the special directories . and .. but in fact ignored all filenames starting in a dot. Oops… But this was (ab?)used and became the standard.

I knew this! Very interesting fact. 

Thank you for all the help thus far!
